File: _src/LinkedList/DoublyLinkedList.ts
type NodeType = DoubleListNode<any> | null;

class DoubleListNode<T> {
  public data: T;
  public next: DoubleListNode<T> | null;
  public prev: DoubleListNode<T> | null;

  constructor(data: T) {
    this.data = data;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList<T> {
  private head: DoubleListNode<T> | null;

  constructor() {
    this.head = null;
  }
}

module.exports = DoublyLinkedList;

File: data_structures/linked_list/doubly_linked_list.py
class ListNode:
    def __init__(self, data):
        self.data = data
        self.next: ListNode = None
        self.prev: ListNode = None


class DoublyLinkedList:
    def __init__(self) -> None:
        self._head: ListNode = None
        self._tail: ListNode = None
        self._length = 0

    def __str__(self) -> str:
        return " <-> ".join(map(str, self.to_list()))

    @property
    def length(self) -> int:
        """
        Returns
            length of the list
        """
        return self._length

    def to_list(self) -> list:
        """
        Returns
            The elements of the LinkedList in a Python list
        """
        r = []
        n = self._head
        while n is not None:
            r.append(n.data)
            n = n.next
        return r
